File: C++/ps_baekjoon/17090/main.cpp
#include <bits/stdc++.h>
using namespace std;

constexpr int mxN = 5e2, mxM = 5e2, out_of_bound = 0, not_visit = -1;
int n, m, cnt, cache[mxN + 2][mxM + 2];
char maze[mxN + 2][mxM + 2];
unordered_map<char, pair<int, int>> direction{{'U', {-1, 0}}, {'R', {0, +1}}, {'D', {+1, 0}}, {'L', {0, -1}}};

bool dfs(int row, int col)
{
    auto &ret = cache[row][col];
    if (maze[row][col] == out_of_bound)
        return true;
    if (ret != not_visit)
        return ret;
    ret = false;
    auto [dr, dc] = direction[maze[row][col]];
    return ret = dfs(row + dr, col + dc);
}

int main()
{
    cin.tie(0)->sync_with_stdio(0);
    memset(cache, not_visit, sizeof(cache));
    cin >> n >> m;
    for (int i = 1; i <= n; ++i)
        for (int j = 1; j <= m; ++j)
            cin >> maze[i][j];
    for (int i = 1; i <= n; ++i)
        for (int j = 1; j <= m; ++j)
            if (dfs(i, j))
                ++cnt;
    cout << cnt;
    return 0;
}

File: Python/ps_baekjoon/17090/main.py
import sys

sys.setrecursionlimit(10 ** 9)
readline, write = (sys.stdin.readline, sys.stdout.write)
direction = {"U": (-1, 0), "R": (0, +1), "D": (+1, 0), "L": (0, -1)}
cnt = 0
out_of_bound = 0
not_visit = -1


def dfs(row, col) -> bool:
    if maze[row][col] == out_of_bound:
        return True
    if cache[row][col] != not_visit:
        return cache[row][col]
    cache[row][col] = False
    dr, dc = direction[maze[row][col]]
    cache[row][col] = dfs(row + dr, col + dc)
    return cache[row][col]


n, m = map(int, readline().strip().split(" "))
maze = [[out_of_bound] * (m + 2) for _ in range(n + 2)]
for i in range(1, n + 1):
    maze[i][1 : m + 1] = [*list(readline().strip())[:m]]
cache = [[not_visit] * (m + 2) for _ in range(n + 2)]
for i in range(1, n + 1):
    for j in range(1, m + 1):
        if dfs(i, j):
            cnt += 1
write(f"{cnt}")